Set variable paths dynamically (#1604)
ImperialCollegeLondon/virtual_rainforest: docs/source/using_the_ve/modifying_a_model.md

Reviewer: Would be good to link to this file from `docs/source/using_the_ve/overview.md` as all the other how to guides/tutorials have links there

---
jupytext:
  formats: md:myst
  main_language: python
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.19.2
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
language_info:
  codemirror_mode:
    name: ipython
    version: 3
  file_extension: .py
  mimetype: text/x-python
  name: python
  nbconvert_exporter: python
  pygments_lexer: ipython3

>   version: 3.11.9

Reviewer 2: Just for curiosity, there is no python code in this md, but these YAML configs are still required for a Jupytext?

Author: Yeah - they aren't required here it is basically just a template we tend to copy between pages 😄

mystnb:
  render_markdown_format: myst
---

# Modifying a model

Once you have set up a model for your site, you are almost certainly going to want to
modify the initial configuration to explore model behaviour under different conditions.

There are a number of different ways you can do this and we describe some options below.
In general, we strongly advise against creating many copies of configuration files and
altering each set. This is hard to do systematically and creates a lot of file
duplication and it is generally easier to use one of the solutions below to run

> alternative models based on the same initial site configuration.

Reviewer 2: ```suggestion
alternative models based on the same initial core set of config files.
```

Suggest to reuse the same phrase as L43.

## Overriding a configuration parameter

> The `ve_run` command accepts a `--config` option that can be used to override one or

Reviewer 2: Should we emphasise the overriding property? E.g., add a line somewhere below like "...`core.constants.max_depth_of_microbial_activity` will override the existing value in the config files"

more of the settings from the configuration file. This makes it fairly easy to have a
single core set of config files and permute a small number of settings.

> For example, say you wanted to rerun the model for your site but using a different

Reviewer 2: ```suggestion
For example, you wanted to rerun the model for your site but using a different
```

maximum soil depth for microbial activity and leaf albedo then you could run:

```bash
ve_run \
  config/data_variables.toml \
  config/soil_config.toml \
  ...

>   config/animal_config.toml \

Reviewer 2: I probably have missed it, but this way of specifying individual config files has always been available? It isn't clear to us, and we thought that one needs to put the files in a single folder and then specify a single path in the args, i.e., we thought inputting individual configs is an upcoming feature. If we knew that then we might have set `maliau_1` and `maliau_2` differently to reduce file duplication. For example, the animal, soil microbial, plant constants, and core.timing are basically the same between these scenarios.

Author: Yes - it has always been available - hence my detailed comments on https://github.com/ImperialCollegeLondon/ve_data_science/pull/321. More on Teams

  --config core.constants.max_depth_of_microbial_activity=0.5 \
  --config abiotic.constants.leaf_albedo=0.2
```

## Using an alternative configuration file

> The approach only works for explicitly named configuration settings and some parts of

Reviewer 2: ```suggestion
The approach above only works for explicitly named configuration settings. However, some parts of
```

I hope I haven't changed the original meaning.

the configuration include tables of entries that do not have direct names. One example
are the data variables: there is no configuration entry
`core.data.variables.air_temperature`.

One way to tackle this problem is to swap one of your configuration files. The Virtual
Ecosystem permits modular configuration files for exactly this reason: it allows you to
build up a library of different configuration options, like a menu, that you can combine
to simulate different situations.

For example, if you wanted to run a simulation under different climate change scenarios,
you might create specific TOML files that only contain the climatic variables and then
have two runs:

```bash
ve_run \
  config/non_climatic_data_variables.toml \
  config/rcp_26_climate_variables.toml \
  config/soil_config.toml \
  ...
  config/animal_config.toml \
```

```bash
ve_run \
  config/non_climatic_data_variables.toml \
  config/rcp_45_climate_variables.toml \
  config/soil_config.toml \
  ...
  config/animal_config.toml \
```

## Iterating over many inputs

If you are trying to explore model sensitivity to parameter variation then it is fairly

> easy to loop over combinations of configuration settings using `--config`. However, this

Reviewer 2: ```suggestion
easy to loop over combinations of explicitly named configuration settings using `--config`. However, this
```

doesn't work if you need to iterate over different input data files. For example, if you
have 1000 simulations of climatic patterns from El Niño-Southern Oscillation events then

> you might want to change the abiotic conditions to use each one of those files.

Reviewer 2: ```suggestion
you might want to change the abiotic conditions to use each one of those input data files.
Below we offer two possible solutions:
```

### File substitution

One approach that works if the models are running in series or on different computers
(such as on an HPC cluster) is to generate a configuration model that points to a
placeholder data file location and then copy the appropriate real data to that location
before running the file.

> For example, your might have a data configuration file `climate_data_variables.toml`

Reviewer 2: ```suggestion
For example, your might have a data configuration file `config/climate_data_variables.toml`
```

This is very not essential, your call!

that looks like this:

```toml
# Climate data
[[core.data.variable]]
file_path = "simulation_inputs/climate.nc"
var_name = "air_temperature_ref"
[[core.data.variable]]
file_path = "simulation_inputs/climate.nc"
var_name = "relative_humidity_ref"
```

And then you copy the correct file into the climate data location before running the
model:

```bash
cp enso_simulations/enso_0001.nc simulation_inputs/climate.nc

ve_run \
  config/climate_data_variables.toml \
  config/soil_config.toml \
  ...
  config/animal_config.toml \
```

### Dynamic file definition

File substitution does not work well if you are running models in parallel on the same
system: the shared location can only point to one file at a time. An alternative
approach is to set the file path dynamically in the configuration using environment
variables.

> In this approach, your climate configuration file `climate_data_variables.toml` would

Reviewer 2: ```suggestion
In this approach, your climate configuration file `config/climate_data_variables.toml` would
```

look like this:

```toml
# Climate data
[[core.data.variable]]

> file_path = "$ENSO_FILE"

Reviewer 2: We need to explain that this is called an "environment variable" to the layperson. I would personally appreciate some hand-holding here. 

Maybe something like:

...where `$ENSO_FILE` is an environment variable that can be defined in the terminal prior to calling `ve_run`. An advantage of setting `$ENSO_FILE` external to `ve_run` is that we can vary it in a bash job array script <blah blah>

var_name = "air_temperature_ref"
[[core.data.variable]]
file_path = "$ENSO_FILE"
var_name = "relative_humidity_ref"
```

You can then set the location of that file before running the model

```bash
export ENSO_FILE=enso_simulations/enso_0001.nc

ve_run \
  config/climate_data_variables.toml \
  config/soil_config.toml \
  ...
  config/animal_config.toml \
```

> ```{important}

Reviewer 2: Sorry I am not sure what this callout box means... Are you trying to explain the scope/limitation of `os.environ.get()` in `virtual_ecosystem/core/configuration.py`?

Author: Yeah - needs more unpacking.
